### PR TITLE
Support for Endpoint URL

### PR DESCRIPTION
**Pull request by nnordrum**
Also cleaned up some deprecated stuff (why the scope is actually so large).

* **Please check if the PR fulfills these requirements**
- [X] The commit message describes your change
- [x] Tests for the changes have been added if possible (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)
- [x] Changes are mentioned in the changelog (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature

* **What is the current behavior?** (You can also link to an open issue here)
endpoint-url is not configurable

* **What is the new behavior (if this is a feature change)?**
endpoint-url _is_  configurable

* **Does this PR introduce a breaking change?** (What changes might users need to make in their setup due to this PR?)
It should not.

* **Other information**:
Changelog? I don't see any documentation on how to  "mention changes in the changelog". I'll add tests tonight/tomorrow. Wanted to get this in early for feedback.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 .idea
 *.iml
 /target
-deploy.sh
\ No newline at end of file
+deploy.sh
+work/*
```
Comment (hoegertn) [inline .gitignore]: ??

**Comment by hoegertn** [comment]
Thanks for your contribution. I really like the original intent of adding endpoint URLs.

What I do not like that much:
* whitespace differences
* non-generic factory method for clients
* s3 config in withAWS step

For changelog entries, just see older entries. Something like "ading endpoitn URL to withAWS step" would be great. please add it to the step documentation too.

File: .gitignore (hunk 1 above)
Comment (nnordrum) [reply .gitignore]: when you start with mvn hpi:run, it creates this work directory. I can't think of a situation where you would want to not ignore this...

Comment (hoegertn) [reply .gitignore]: Ok sounds, great. Did not do this before.

**Comment by nnordrum** [comment]
What I do not like that much:

- whitespace differences
  - I just saved them using the default settings in IntelliJ. I guess I can go back and adjust whitespace if it's really that big of a deal...
- non-generic factory method for clients
  - While I generally agree with you, [that's not how the AWS SDK works now](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/client/builder/AwsSyncClientBuilder.html). I guess I could put the generic one back in and then delegate the creation to the sub methods, or if there is some generic builder that I'm just not aware of, let me know and I'll switch it back. 
- s3 config in withAWS step
  - So there are a couple options that I saw here:
    1. put them in `withAWS`
        - Pros: centralized configuration, parameters are optional, configure once for multiple S3 Steps
        - Con: "extra" options when using the Pipeline Snippet Generator
    1. create a `withAWSS3`
        - Pros: _scoped_ centralized configuration, parameters are optional, configure once for multiple S3 Steps
        - Con: extra sub-closure
    1. put them individually in the `S3*Step` classes
        - Pros: config options are specific to the `S3*Step`
        - Con: lots of duplicated configuration if calling `S3*Step`s multiple times

IMO, `c` is clearly the worst of the 3 options.

I think _ideally_, if you want the most flexibility, you would allow configuration at all 3 levels, but I think that's probably overkill.

I went with `A` over `B` due to simplicity. I also intentionally kept the name _exactly_ the same as the method in the Client Builder in the AWS SDK so it would be clear how it was being used.

I think moving to `B` in the future might be a good idea, but I could see it getting _really_ ugly as you have
``` groovy
withAWS(endpoint-url: "") {
  withS3(someS3Param: "", anotherS3Param: "") {
    withLambda(someLambdaParam: "", anotherLambdaParam: "") {
      withSNS(someSNSParam: "", anotherSNSParam: "") {
        withACR(someACRParam: "", anotherACRParam: "") {
          /* do lots of stuff */
        }
      }
    }
  }
}
```
defining service-specific parameters. I, personally, would rather see this:
``` groovy
withAWS(endpoint-url: ""
          , someS3Param: "", anotherS3Param: ""
          , someLambdaParam: "", anotherLambdaParam: ""
          , someSNSParam: "", anotherSNSParam: ""
          , someACRParam: "", anotherACRParam: "") {
  /* do lots of stuff */
}
```


If `withService` intermediaries are defined, I think allowing options to be defined at the `withAWS` _and_ the `withService` scopes would be very helpful, but all of that is _way_ outside the scope of this PR :smile:

**Comment by hoegertn** [comment]
Hi,

for the generic factory:
What about a create function that takes the builder like:
```
client = create(AmazonCloudFormationClientBuilder.standard(), vars);
```

for the s3Config:
what do you think about splitting the PR so we can get the endpoint stuff in and can discuss the s3Params separately?

**Comment by nnordrum** [comment]
Technically yes, but for me they are actually tied together, since I’m using endpoint-url to set to an internally hosted S3-compatible server that is setup to use path access only at this time.

**Comment by hoegertn** [comment]
After careful thought, I think I would go for option `c` for s3Config because it does have to minimal impact on other parts of the plugin.

Combined with my proposal for builder-style factory methods, you can then provide the pre-configured builder to the factory.

**Comment by nnordrum** [comment]
rebuild

**Comment by nnordrum** [comment]
@hoegertn any idea when you’re planning to do the next release? No rush, just curious. We’re already running my PR branch locally anyway ;)

**Comment by hoegertn** [comment]
I plan to address some bugs and do a release at the end of the week.

**Pull request merged**